lib/OpenQA/Worker/Common.pm

```diff
@@ -138,43 +145,83 @@ sub api_init {
     }
 }
 
+my $call_running;
 # send a command to openQA API
 sub api_call {
     my ($method, $path, $params, $json_data, $ignore_errors) = @_;
-    $method = lc $method;
+
+    return undef unless verify_workerid();
+    if ($call_running) {
+        print "api call in progress, not accepting new one\n" if $verbose;
+        return undef;
+    }
+
+    $call_running = 1;
+
+    $method = uc $method;
     my $ua_url = $url->clone;
 
     $ua_url->path($path =~ s/^\///r);
     $ua_url->query($params) if $params;
 
-    my $tries = 3;
-    while (1) {
-        print uc($method) . " $ua_url\n" if $verbose;
-        my $tx;
-        if ($json_data) {
-            $tx = $ua->$method($ua_url => json => $json_data);
-        }
-        else {
-            $tx = $ua->$method($ua_url);
-        }
+    print $method . " $ua_url\n" if $verbose;
+
+    my @args = ($method, $ua_url);
+
+    if ($json_data) {
+        push @args, 'json', $json_data;
+    }
+
+    my $res;
+    my $done = 0;
+
+    my $tx = $ua->build_tx(@args);
+    my $cb;
+    $cb = sub {
+        my ($ua, $tx, $tries) = @_;
+        remove_timer('api_call');
         if ($tx->success) {
-            return $tx->success->json;
+            $res = $tx->success->json;
+            $done = 1;
+            return;
         }
         elsif ($ignore_errors) {
+            $done = 1;
             return;
         }
         --$tries;
-        my ($err, $code) = $tx->error;
-        my $msg = $code ? "$tries: $code response: $err" : "$tries: Connection error: $err->{message}";
-        carp "$msg";
+        my $err = $tx->error;
+        my $msg = $err->{code} ? "$tries: $err->{code} response: $err->{message}" : "$tries: Connection error: $err->{message}";
+        warn $msg;
         if (!$tries) {
             # abort the current job, we're in trouble - but keep running to grab the next
-            # this lives in Jobs.pm - this is recursive use?
-            #stop_job('api-failure');
+            remove_timer('setup_websocket');
+            $workerid = undef;
+            OpenQA::Worker::Jobs::stop_job('api-failure');
+            add_timer('register_worker', 10, \&register_worker, 1);
+            $done = 1;
             return;
         }
-        sleep 5;
+        $tx = $ua->build_tx(@args);
+        add_timer(
+            'api_call',
+            5,
+            sub {
+                $ua->start($tx => sub { $cb->(@_, $tries) });
+            },
+            1
+        );
+    };
+    $ua->start($tx => sub { $cb->(@_, 3) });
+
+    # This ugly. we need to "block" here so enter ioloop recursively
+    while(!$done && Mojo::IOLoop->is_running) {
```

Comment:
Unfortunately delay->wait only actually waits if there is no even loop yet. It just continues if there is already running. However it may actually work to use a delay to chain three times the callback and somehow pass the result between them. hmm. might be worth a try in the future.